sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/OptimizeSortMergeJoinWithPartialHashDistribution.scala

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.exchange
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.physical.{HashPartitioning, Partitioning}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{SortExec, SparkPlan}
+import org.apache.spark.sql.execution.joins.SortMergeJoinExec
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This rule removes shuffle for the sort merge join if the following conditions are met:
+ * - The child of ShuffleExchangeExec has HashPartitioning with the same number of partitions
+ *   as the other side of join.
+ * - The child of ShuffleExchangeExec has output partitioning which has the subset of
+ *   join keys on the respective join side.
+ *
+ * If the above conditions are met, shuffle can be eliminated for the sort merge join
+ * because rows are sorted before join logic is applied.
+ */
+case class OptimizeSortMergeJoinWithPartialHashDistribution(conf: SQLConf) extends Rule[SparkPlan] {
+  def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.optimizeSortMergeJoinWithPartialHashDistribution) {
+      return plan
+    }
+
+    plan.transformUp {
+      case s @ SortMergeJoinExec(_, _, _, _,
+        lSort @ SortExec(_, _,
+          ExtractShuffleExchangeExecChild(
+            lChild,
+            lChildOutputPartitioning: HashPartitioning),
+          _),
+        rSort @ SortExec(_, _,
+          ExtractShuffleExchangeExecChild(
+            rChild,
+            rChildOutputPartitioning: HashPartitioning),
+          _),
+        false) if isPartialHashDistribution(
+          s.leftKeys, lChildOutputPartitioning, s.rightKeys, rChildOutputPartitioning) =>
+        // Remove ShuffleExchangeExec.
+        s.copy(left = lSort.copy(child = lChild), right = rSort.copy(child = rChild))
+      case other => other
+    }
+  }
+
+  /*
+   * Returns true if both HashPartitioning have the same number of partitions and
+   * their partitioning expressions are a subset of their respective join keys.
+   */
+  private def isPartialHashDistribution(
+      leftKeys: Seq[Expression],
+      leftPartitioning: HashPartitioning,
+      rightKeys: Seq[Expression],
+      rightPartitioning: HashPartitioning): Boolean = {
+    val mapping = leftKeyToRightKeyMapping(leftKeys, rightKeys)
+    (leftPartitioning.numPartitions == rightPartitioning.numPartitions) &&
+      leftPartitioning.expressions.zip(rightPartitioning.expressions)
+        .forall {
+          case (le, re) => mapping.get(le.canonicalized)
+            .map(_.exists(_.semanticEquals(re)))
+            .getOrElse(false)
+        }
```

Comment:
Thanks. I agree with your concerns for both cases. But, for the first example, only one side will be shuffled, so the rule should not kick in. For the second example,  we have `t1.a = t2.b AND t1.b = t2.a` which matches the bucket ordering, so this should be also fine.

Reply:
Sorry if I miss anything:

> But, for the first example, only one side will be shuffled, so the rule should not kick in.

If the number of buckets for `t1` is less than number of shuffle partitions, shouldn't it shuffle both sides ? (in [`EnsureRequirements`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala#L96)). So the rule kicks in here and removes both shuffles, but we shouldn't remove any shuffle here.

> For the second example, we have t1.a = t2.b AND t1.b = t2.a which matches the bucket ordering, so this should be also fine.

I think it's unsafe if we do not shuffle both sides. `HashPartitioning(Seq(a, b))` and `HashPartitioning(Seq(b, a))` are not same thing, e.g. for tuple (a: 1, b: 2) it will be assigned to different buckets given current `Murmur3Hash` implementation.

Reply:
> If the number of buckets for `t1` is less than number of shuffle partitions, shouldn't it shuffle both sides ? (in [`EnsureRequirements`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala#L96)). So the rule kicks in here and removes both shuffles, but we shouldn't remove any shuffle here.

You are right. Thanks for the catch!

> I think it's unsafe if we do not shuffle both sides. `HashPartitioning(Seq(a, b))` and `HashPartitioning(Seq(b, a))` are not same thing, e.g. for tuple (a: 1, b: 2) it will be assigned to different buckets given current `Murmur3Hash` implementation.

Yes, I understand they produce different hash values. However, it has the join condition `t1.a = t2.b AND t1.b = t2.a`. This rule will not be applied if the condition was `t1.a = t2.a AND t1.b = t2.b`. Please let me know if I missed something. Thanks!